In [17]:
# Imports and pip installations (if needed)


import pandas as pd

# Part 1: Load the dataset

In [18]:
# Load the given datasets


import pandas as pd
numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
print(numerical)
categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
print(categorical)
# Print the data


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
Each category within the data represents different information about a patient within the kidney disease database. The data gives insight about levels of vitamins/nutrients/body necessities within the person. The two datasets both share one category of "unique_id" which is a way to identify each patient. Other than that the datasets split into numerical and categorical data, with the numerical data being focused on measurable levels of these necessary body nutrients, and the categorical data focused on a Yes/No or High/Med/Low.
Examples of Numerical data include:
Age: represents a patients age in years
bp: represents a patients Blood pressure in mm/Hg (Millimeters of mercury)
sod: represents a patient Sodium level in mEq/L (milliequivalents per liter)

Examples of Categorical data include:
cad: represents if somebody has coronary artery disease, yes or no
ane: represents if somebody has anemia, yes or no
dm: Represents if somebody has diabetes mellitus, yes or no

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [19]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
#print(len(numerical))
duplicate_rows_num = numerical[numerical.duplicated(subset="unique_id")]
if not duplicate_rows_num.empty:
    print(len(duplicate_rows_num)) #total rows with a duplicate are 59, remove 30 of them
    numerical = numerical.drop_duplicates(subset="unique_id")
#print(len(numerical))
# Drop these duplicate rows


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
#print(len(categorical))
duplicate_rows_cat = categorical[categorical.duplicated(subset="unique_id")]
if not duplicate_rows_cat.empty:
    print(len(duplicate_rows_cat)) #total rows with a duplicate are 59, remove 30 of them
    categorical = categorical.drop_duplicates(subset="unique_id")
#print(len(categorical))


30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [20]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined = pd.merge(numerical, categorical, on="unique_id", how="outer")
#Print the combined dataset
print(combined)



     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [21]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = len(combined)
Na_rows = combined.isnull().any(axis=1).sum()
print(round(((Na_rows/total_rows)*100),2))
# Print %

# Drop these rows from the dataset

cleaned = combined.dropna()
print(cleaned)
# Print the Dataset

61.75
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   

## Part 3.4: Sort the dataset according to the Labels

In [22]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted = cleaned.sort_values(by="Target")
sorted = sorted.reset_index(drop=True)
# Print the dataset
print(sorted)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [23]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dummy_cats = pd.get_dummies(sorted, columns=categorical_columns, drop_first=True, dtype=int)
print(dummy_cats)
# Print the dataset

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [24]:
# Remove outliers
#print(len(dummy_cats))

numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    mean = dummy_cats[col].mean()
    std = dummy_cats[col].std()
    upperB = mean + 3*std
    lowerB = mean - 3*std

    dummy_cats = dummy_cats[(dummy_cats[col] <= upperB) & (dummy_cats[col] >= lowerB)]
dummy_cats = dummy_cats.reset_index(drop=True)    
print(dummy_cats)

# Print the dataset

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
128     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
129     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
130     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
131     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
132     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [25]:
# Normalize the all Numerical Attributes in the dataset.
dummy_cats[numerical_columns] = dummy_cats[numerical_columns].apply(lambda x: ((x - x.mean())/x.std()))
    
print(dummy_cats)
#for col in dummy_cats:
    #print(col)
# Print the dataset

     unique_id       age        bp       bgr        bu        sc       sod  \
0       343710  0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087   
1       397388  0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053   
2       438182  1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910   
3       474407 -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013   
4       475200  0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124   
..         ...       ...       ...       ...       ...       ...       ...   
128     486397 -1.769750 -0.291121  0.036896  0.111307 -0.264117 -0.956087   
129     203466 -1.644048 -1.310043 -0.537820 -0.860662 -0.129786 -0.429910   
130     482105 -0.261328 -1.310043 -0.582029  0.186074 -0.599945 -0.429910   
131     228849  0.618586  0.727802 -0.471507  0.335608 -0.129786 -0.079124   
132     996076  0.555735 -1.310043 -0.360984  0.298224 -0.129786  1.674801   

          pot      hemo       pcv  ...  pc_normal  pcc_present 

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The scientists are trying to use this data to find a cure for Chronic Kidney Disease (CKD). With than being said its necessary to keep all of the patient information. The only column that is not necessary is unique_id. At this point identifying the specific patient is unnecessary since they have all the information already.  

In [26]:
#Remove that column
finalized_data = dummy_cats.drop(columns="unique_id")

# Print the dataset
print(finalized_data)

          age        bp       bgr        bu        sc       sod       pot  \
0    0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087  1.367853   
1    0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053 -1.298281   
2    1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910  2.308841   
3   -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013 -1.298281   
4    0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124  0.426864   
..        ...       ...       ...       ...       ...       ...       ...   
128 -1.769750 -0.291121  0.036896  0.111307 -0.264117 -0.956087 -0.827787   
129 -1.644048 -1.310043 -0.537820 -0.860662 -0.129786 -0.429910 -0.043630   
130 -0.261328 -1.310043 -0.582029  0.186074 -0.599945 -0.429910 -0.200461   
131  0.618586  0.727802 -0.471507  0.335608 -0.129786 -0.079124 -1.298281   
132  0.555735 -1.310043 -0.360984  0.298224 -0.129786  1.674801  0.583696   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present 

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [27]:
# Export the dataframe to a new csv file
finalized_data.to_csv('finalized_data.csv', index=False)

# Export the dataframe to a new json file
finalized_data.to_json('finalized_data.json', orient='records')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [28]:
#print(finalized_data.head(15))

### Paste the markdown table here

|    |     age   |     bp   |    bgr   |    bu    |    sc    |    sod   |    pot   |    hemo    |     pcv    |     wbcc   |  pc_normal |  pcc_present |  ba_present |  htn_yes |  dm_yes |  cad_yes |  appet_poor |  pe_yes |  ane_yes |  Target_notckd |
|----|-----------|----------|----------|----------|----------|----------|----------|------------|------------|------------|------------|--------------|-------------|----------|---------|----------|-------------|---------|----------|----------------|
|  0 |  0.807138 | 0.727802 | 1.142120 | 3.999185 | 1.683685 | -0.956087|  1.367853 |  -2.902248 |  -2.894560 |  0.517701  |  0         |  0           |  0          |  1       |  1      |  1       |  1          |  1      |  1       |  0              |
|  1 |  0.932840 | 2.765646 | 3.507298 | -0.225144| 1.213526 |  0.447053| -1.298281 |  -0.604893 |  -0.588449 |  0.782525  |  1         |  0           |  1          |  1       |  0      |  1       |  0          |  0      |  0       |  0              |
|  2 |  1.561350 | 0.727802 | 2.910477 | 3.774884 | 2.153844 | -0.429910|  2.308841 |  -1.688551 |  -1.597372 | -0.365045  |  0         |  0           |  0          |  1       |  1      |  1       |  0          |  0      |  0       |  0              |
|  3 | -1.706899 | 1.746724 | -0.316775| -0.038227| 0.206042 | -2.710013| -1.298281 |  -2.642170 |  -3.038692 |  1.930095  |  0         |  1           |  1          |  0       |  0      |  0       |  0          |  0      |  1       |  0              |
|  4 |  0.995691 | 2.765646 | 0.921075 | 0.485141 | 3.900150 | -0.079124|  0.426864 |  -2.815556 |  -2.606296 | -0.232633  |  0         |  0           |  1          |  1       |  1      |  0       |  0          |  1      |  0       |  0              |
|  5 |  0.744287 | -1.31004 | 3.684134 | -0.187760| 0.206042 | -1.833050| -2.082438 |  -2.815556 |  -2.750428 |  3.165939  |  0         |  1           |  0          |  1       |  0      |  0       |  1          |  0      |  1       |  0              |
|  6 |  0.492884 | 1.746724 | 0.169523 | 2.466464 | 3.564322 | -1.657657|  0.740527 |  -2.295400 |  -2.173900 | -0.718144  |  0         |  0           |  0          |  1       |  0      |  0       |  0          |  0      |  0       |  0              |
|  7 |  1.309946 | -0.29112 | 2.048403 | 2.055246 | 3.295659 | -3.586975| -0.670955 |  -2.165361 |  -2.318032 |  1.532859  |  0         |  1           |  1          |  1       |  1      |  1       |  0          |  1      |  1       |  0              |
|  8 | -0.135626 | -1.31004 | 0.921075 | 1.905712 | 1.280691 |  0.096268| -0.514124 |  -1.991975 |  -2.318032 |  2.901115  |  1         |  0           |  0          |  1       |  1      |  0       |  0          |  0      |  0       |  0              |
|  9 |  0.869989 | -0.29112 | 0.014792 | 0.036540 | 0.206042 | -0.780695|  0.583696 |  -0.778278 |  -0.732580 | -0.056084  |  0         |  0           |  0          |  1       |  1      |  0       |  0          |  0      |  0       |  0              |
| 10 |  0.430033 | 0.727802 | 2.048403 | 1.195427 | 1.683685 | -0.605302|  0.897359 |  -1.515166 |  -1.453240 | -0.276771  |  0         |  1           |  1          |  1       |  1      |  0       |  0          |  1      |  0       |  0              |
| 11 |  0.995691 | -1.31004 | 2.601015 | 0.634675 | 1.952347 | -0.605302|  1.681516 |  -2.122014 |  -2.173900 | -0.232633  |  0         |  0           |  1          |  1       |  1      |  0       |  1          |  1      |  0       |  0              |
| 12 |  1.435648 | -1.31004 | -0.07362 | 3.139366 | 2.624003 | -0.780695|  0.897359 |  -1.298434 |  -1.309108 |  3.165939  |  1         |  0           |  0          |  1       |  1      |  0       |  1          |  1      |  0       |  0              |
| 13 |  1.435648 | -0.29112 | 2.158925 | 1.531878 | 1.482188 | -1.306872|  0.113202 |  -1.731897 |  -1.597372 | -1.071242  |  0         |  1           |  1          |  1       |  1      |  1       |  0          |  0      |  0       |  0              |
| 14 |  0.681437 | 0.727802 | 4.015701 | -0.225144| -0.062621| -3.236190| -1.298281 |  -1.731897 |  -1.309108 |  1.268035  |  1         |  0           |  0          |  0       |  1      |  0       |  0          |  0      |  0       |  0              |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Besides the obvious of cross checking the data and the output of a LLM conversion, you could have the LLM do the conversions in chunks to minimize the amount of data that it has to do at once. Lastly you could have the LLM do the conversions in steps first. This minimizes the amount of hallucinations because it only focuses on one task at a time

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [29]:
# Code to rename all the columns in the dataset
column_renames = {
    'age': 'age_years',
    'bp': 'blood_pressure_mmHg',
    'sg': 'specific_gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red_blood_cells',
    'pc': 'pus_cells',
    'pcc': 'pus_cell_clumps',
    'ba': 'bacteria',
    'bgr': 'blood_glucose_random_mg_dl',
    'bu': 'blood_urea_mg_dl',
    'sc': 'serum_creatinine_mg_dl',
    'sod': 'sodium_mEq_L',
    'pot': 'potassium_mEq_L',
    'hemo': 'hemoglobin_gm',
    'pcv': 'packed_cell_volume',
    'wbcc': 'white_blood_cell_count_cells_cmm',
    'rbcc': 'red_blood_cell_count_millions_cmm',
    'htn': 'hypertension',
    'dm': 'diabetes_mellitus',
    'cad': 'coronary_artery_disease',
    'appet': 'appetite',
    'pe': 'pedal_edema',
    'ane': 'anemia',
    'class': 'ckd_classification'
}

toReturn = finalized_data.copy()
toReturn.rename(columns=column_renames, inplace=True)


print(toReturn)

     age_years  blood_pressure_mmHg  blood_glucose_random_mg_dl  \
0     0.807138             0.727802                    1.142120   
1     0.932840             2.765646                    3.507298   
2     1.561350             0.727802                    2.910477   
3    -1.706899             1.746724                   -0.316775   
4     0.995691             2.765646                    0.921075   
..         ...                  ...                         ...   
128  -1.769750            -0.291121                    0.036896   
129  -1.644048            -1.310043                   -0.537820   
130  -0.261328            -1.310043                   -0.582029   
131   0.618586             0.727802                   -0.471507   
132   0.555735            -1.310043                   -0.360984   

     blood_urea_mg_dl  serum_creatinine_mg_dl  sodium_mEq_L  potassium_mEq_L  \
0            3.999185                1.683685     -0.956087         1.367853   
1           -0.225144              

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [30]:
# Converted SQL to Pandas code
result = finalized_data.groupby('Target_notckd').size().reset_index(name='count')
print(result)

   Target_notckd  count
0              0     22
1              1    111


*Describe what the above code does here*

finalized_data.groupby('Target_notckd') is similar to the GROUP BY Target clause in SQL. 
It groups the rows in the DataFrame/ SQL table by the unique values in the Target column.

.size() corresponds to the COUNT(*) function in SQL. 
It calculates the number of rows  in each group created by groupby/GROUP BY in SQL.

.reset_index(name='count') is like adding a SELECT Target, COUNT(*) AS count clause. 
It converts the grouped results into a regular DataFrame/SQL set, and it names the count column as count

Overall, the SQL query that groups data by the Target/Target_notckd column and counts occurrences is translated into pandas using groupby('Target_notckd'), followed by .size() to count rows in each group, and .reset_index() to format the result as a new DataFrame with the count column.
